### PR TITLE
Add support for held item and off-hand swap

### PR DESCRIPTION
**Pull request by Sceat**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -107,6 +107,7 @@ const initial_state = {
   inventory_sequence_number: 0,
   inventory_cursor: null,
   inventory_cursor_index: 0,
+  inventory_held_slot_index: 0, // goes from 0 to 8
```
Comment (DeltaEvo) [inline src/index.js:110]: Don't want to describe that with a type :smile:  ?
```suggestion
  /** @type {0|1|2|3|4|5|6|7|8} */ 
  inventory_held_slot_index: 0,
```

File: src/player/inventory.js [hunk 2]
```diff
@@ -25,50 +25,84 @@ const BlockDigStatus = {
   SWAP_ITEM_IN_HAND: 6,
 }
 
+const OFF_HAND_SLOT = 45
```
Comment (DeltaEvo) [inline src/player/inventory.js:28]: Can you move your code into a new file instead, like src/player/held_item.js ?

File: src/player/inventory.js [hunk 3]
```diff
@@ -25,50 +25,84 @@ const BlockDigStatus = {
...
+      state.inventory_held_slot_index
+    )
+    ;[state.inventory[OFF_HAND_SLOT], state.inventory[held_item_slot]] = [
+      state.inventory.at(held_item_slot),
+      state.inventory.at(OFF_HAND_SLOT),
+    ]
+    return {
+      ...state,
+      inventory: state.inventory,
+      inventory_sequence_number: state.inventory_sequence_number + 1,
```
Comment (DeltaEvo) [inline src/player/inventory.js:82]: We don't need to resync the inventory here

File: src/player/inventory.js [hunk 4]
```diff
@@ -25,50 +25,84 @@ const BlockDigStatus = {
...
+
+  // swap second hand
+  if (status === 6) {
+    const held_item_slot = hotbar_to_inventory_slot(
+      state.inventory_held_slot_index
+    )
+    ;[state.inventory[OFF_HAND_SLOT], state.inventory[held_item_slot]] = [
+      state.inventory.at(held_item_slot),
+      state.inventory.at(OFF_HAND_SLOT),
+    ]
```
Comment (DeltaEvo) [inline src/player/inventory.js:78]: Don't mutate the state
Don't use at if you don't have negative indexes

File: src/index.js (hunk 1 above)
Comment (DeltaEvo) [reply src/index.js:110]: I think we should remove the `inventory_` prefix here

**Review by DeltaEvo — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 1 above)
Comment (Sceat) [reply src/index.js:110]: > Don't want to describe that with a type 😄 ?

image

**Comment by DeltaEvo** [comment]
Mmh it seems that you removed the code that was writing the `held_item_slot`

**Comment by Sceat** [comment]
fixed ✔️ 
and actually we needed the inventory sequence incrementation

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**